awx/main/scheduler/dag_workflow.py

```diff
@@ -215,10 +214,19 @@ def mark_dnr_nodes(self):
             obj = node['node_object']
 
             if obj.do_not_run is False and not obj.job and node not in root_nodes:
-                parent_nodes = [p['node_object'] for p in self.get_dependents(obj)]
+                parent_nodes = [p['node_object'] for p in self.get_parents(obj)]
                 if self._are_all_nodes_dnr_decided(parent_nodes):
                     if self._should_mark_node_dnr(node, parent_nodes):
                         obj.do_not_run = True
                         nodes_marked_do_not_run.append(node)
+            if obj.all_parents_must_converge:
+                if self._are_relevant_parents_finished(node):
+                    # if the current node is a convergence node and all the 
+                    # parents are finished then check to see if all parents
+                    # met their success criteria to run the convergence child
+                    parent_nodes = [p['node_object'] for p in self.get_parents(obj)]
+                    if not all(obj in self.get_children(p, p.job.status) for p in parent_nodes):
```

Comment:
I don't see how this checks the parent's success criteria, as the comment says.

Maybe `get_children` only returns the nodes that correspond to the parent's status (successful / failure), but it would be good to make that more clear. But any DNR parent would make the node DNR in this case right?

It's also not really right to say that it's the parent's "success criteria" in play here. If the parent node is failed, and the node is one of its failure nodes, it can still run, right?

Reply:
`get_children` takes a node and a status, so in this case the parent node and the status. If it were failure it would be all 'failure' child nodes.

The all() leverages that to iterate through all parents of the convergence node and checks that the child (i.e. the convergence node) is in the listed child nodes that match the parents job status. (following the failure example from above, if it failed it will look that the convergence node is in the failure pathway for the parent, if it is not it knows that it should not run it, because convergence was not met).

Also yes you are correct, I'm using the term "success criteria" but am definitely looking for better terminology. I plan on rewriting the comment anyways, because your question makes it clear that it's a bit confusing. :smile: any suggestions on a replacement for "success criteria"?

Reply:
That still doesn't handle the case where a parent is marked DNR, right? You could have a successful parent, and be in the success_nodes relationship. That would passes these tests, but if the parent node was marked DNR, and the child has this `all_parents_must_converge` field set to True, then that parent never ran so you shouldn't count it as converged. I don't see any mention of this case skimming the issue, but it needs to be handled.

Reply:
if a parent is marked as DNR it won't have a status to compare it's child relationships to, meaning it won't meet convergence behavior, so I do think the existing code _could_ cover that case. However, it does bring up a good question about the `get_children` function... if a parent is marked as DNR, will it have a status or not (i.e. will it break the `get_children` query)? I'll test a bit and then update.

Reply:
it definitely broke `get_children` so I added in a check for that.
I've tested the new code and handles DNR parents (although tbh I don't think the code is quite as pretty as before :woman_shrugging: )